### PR TITLE
Use net/context to properly cancel kex2 when prompt canceled

### PR DESCRIPTION
**Pull request by patrickxb**
r? @maxtaco

### CODE REVIEW

File: go/engine/device_add.go [hunk 1]
```diff
@@ -100,7 +99,8 @@ func (e *DeviceAdd) Run(ctx *Context) (err error) {
 			e.G().Log.Debug("received secret phrase, adding to provisioner")
 			ks, err := libkb.NewKex2SecretFromPhrase(receivedSecret.Phrase)
 			if err != nil {
-				e.G().Log.Warning("DisplayAndPromptSecret error: %s", err)
+				e.G().Log.Warning("NewKex2SecretFromPhrase error: %s", err)
+				canceler()
```
Comment (maxtaco) [inline go/engine/device_add.go]: Why is the `defer func() { canceler() }` on line 110 still needed? And why wasn't it sufficient?

File: go/engine/kex2_provisioner.go [hunk 2]
```diff
@@ -102,9 +102,15 @@ func (e *Kex2Provisioner) Run(ctx *Context) error {
...
+	nctx := ctx.NetContext
+	if nctx == nil {
+		e.G().Log.Debug("no NetContext in engine.Context, using context.TODO")
+		nctx = context.TODO()
+	}
+
 	// all set:  start provisioner
 	karg := kex2.KexBaseArg{
-		Ctx:           context.TODO(),
+		Ctx:           nctx,
```
Comment (patrickxb) [inline go/engine/kex2_provisioner.go]: see here...it had context.TODO().

File: go/engine/device_add.go (hunk 1 above)
Comment (maxtaco) [inline go/engine/device_add.go]: Is it ok to call `cancel` twice?  Should we get rid of the `if canceler != nil` mechanism below?

File: go/engine/kex2_provisioner.go [hunk 3]
```diff
@@ -102,9 +102,15 @@ func (e *Kex2Provisioner) Run(ctx *Context) error {
 
 	deviceID := e.G().Env.GetDeviceID()
 
+	nctx := ctx.NetContext
+	if nctx == nil {
+		e.G().Log.Debug("no NetContext in engine.Context, using context.TODO")
+		nctx = context.TODO()
```
Comment (maxtaco) [inline go/engine/kex2_provisioner.go]: should we change this to be `context.Background()` since it's already 'done'?

**Comment by maxtaco** [comment]
ok a few follow-ons...

File: go/engine/device_add.go (hunk 1 above)
Comment (patrickxb) [inline go/engine/device_add.go]: yes, you can call `cancel` repeatedly.

https://github.com/golang/net/blob/master/context/pre_go17.go#L170

i'll remove conditional

**Comment by patrickxb** [comment]
ok, PTAL

**Comment by maxtaco** [comment]
👍

**Pull request merged**